macrosynergy/panel/basket_performance_class.py

```diff
@@ -0,0 +1,390 @@
+            dfwa *= np.sqrt(252)
+
+        else:
+
+            half_life = lback_periods
+            weights = expo_weights(lback_periods * 2, half_life)
+            dfwa = dfw_ret.rolling(window=lback_periods * 2).agg(expo_std, w=weights,
+                                                                 remove_zeros=remove_zeros)
+
+        df_isd = 1 / dfwa
+        df_wgts = df_isd / df_isd.sum(axis=1).values[:, np.newaxis]
+        self.check_weights(df_wgts)
+
+        return df_wgts
+
+    def values_weight(self, dfw_ret: pd.DataFrame, dfw_wgt: pd.DataFrame,
+                      weight_meth: str):
+        """
+        Returns weights based on an external weighting category.
+
+        :param <pd.DataFrame> dfw_ret: Standard wide dataframe of returns across time and
+            contracts.
+        :param <pd.DataFrame> dfw_wgt: Standard wide dataframe of weight category values
+            across time and contracts.
+        :param <str> weight_meth:
+
+        :return <pd.DataFrame>: Dataframe of weights.
+        """
+
+        negative_condition = np.any((dfw_wgt < 0).to_numpy())
+        if negative_condition:
+            dfw_wgt[dfw_wgt < 0] = 0.0
+            print("Negative values in the weight matrix set to zero.")
+
+        exo_array = dfw_wgt.to_numpy()
+        df_bool = ~dfw_ret.isnull()
+
+        weights_df = df_bool.multiply(exo_array)
+        cols = weights_df.columns
+
+        # Zeroes treated as NaNs.
+        weights_df[cols] = weights_df[cols].replace({'0': np.nan, 0: np.nan})
+
+        if weight_meth != "values":
+            weights_df = 1 / weights_df
+
+        weights = weights_df.divide(weights_df.sum(axis=1), axis=0)
+        self.check_weights(weights)
+
+        return weights
+
+    @staticmethod
+    def max_weight_func(weights: pd.DataFrame, max_weight: float):
+        """
+        Enforces maximum weight caps or - if impossible applies equal weight.
+
+        :param <pd.DataFrame> weights: Corresponding weight matrix. Multidimensional.
+        :param <float> max_weight: Upper-bound on the weight allowed for each
+            cross-section.
+
+        :return <pd.DataFrame>: Will return the modified weight DataFrame.
+
+        N.B.: If the maximum weight is less than the equal weight weight, this replaces
+        the computed weight with the equal weight. For instance,
+        [np.nan, 0.63, np.nan, np.nan, 0.27] becomes [np.nan, 0.5, np.nan, np.nan, 0.5].
+        Otherwise, the function calls the ConvergeRow Class to ensure all weights
+        "converge" to a value within the upper-bound. Allow for a margin of error set to
+        0.001.
+        """
+
+        dfw_wgs = weights.to_numpy()
+
+        for i, row in enumerate(dfw_wgs):
+            row = ConvergeRow.application(row, max_weight)
+            weights.iloc[i, :] = row
+
+        return weights
+
+    def weight_dateframe(self, weight_meth: str = "equal", weights: List[float] = None,
```

Comment:
This is a class method whose output is not used in the class. That is good for transparency, but the smoother way would be to have [1] a method `make_weights` that calculates and stores the weights and allows using them in the `performance` method or visualizations and [2] a method `return_weights` that allows returning the weight df for external use, such as target positions.